cmp/chart/impl/ChartContextMenuItems.ts

```diff
@@ -0,0 +1,201 @@
+    | 'downloadJPEG'
+    | 'downloadPNG'
+    | 'downloadSVG'
+    | 'downloadCSV'
+    | 'downloadXLS'
+    | 'downloadPDF'
+    | 'copyToClipboard';
+
+export interface ChartMenuItem extends Omit<MenuItem, 'actionFn' | 'items'> {
+    items?: ChartMenuItemLike[];
+    actionFn?: (
+        menuItemEvent: MouseEvent | PointerEvent,
+        contextMenuEvent: MouseEvent | PointerEvent,
+        params: {
+            chartModel: ChartModel;
+            /**
+             * Single point is the active series point the mouse is closest to
+             */
+            point: any;
+            /**
+             * Points array is the list of points hovered over in each series. When
+             * there are multiple series and tooltip.shared = true, points.length > 1.
+             */
+            points: any[];
+        }
+    ) => void;
+}
+
+/**
+ * If a String, value can be '-' for a separator, or a token supported by HighCharts
+ * for its native menu items, or a Hoist specific token.
+ */
+export type ChartContextMenuItemLike = ChartMenuItem | ChartContextMenuToken | string;
+
+/**
+ * Specification for a ChartContextMenu.  Either a list of items or a function to produce one.
+ */
+export type ChartContextMenuSpec =
+    | boolean
+    | ChartContextMenuItemLike[]
+    | ((chartModel: ChartModel) => ChartContextMenuItemLike[]);
+
+export function getChartContextMenuItems(
+    items: ChartContextMenuItemLike[],
+    contextMenuEvent: MouseEvent | PointerEvent,
+    chartModel: ChartModel,
+    point,
+    points
+) {
+    return cloneDeep(items).map(it =>
+        buildMenuItemConfig(it, contextMenuEvent, chartModel, point, points)
+    );
+}
+
+//---------------------------
+// Implementation
+//---------------------------
+function buildMenuItemConfig(
+    item: ChartContextMenuItemLike,
+    contextMenuEvent: MouseEvent | PointerEvent,
+    chartModel: ChartModel,
+    point,
+    points
+): MenuItem | string {
+    if (isString(item)) return parseToken(item, chartModel);
+
+    // build nested menu item configs
+    if (isMenuItem(item)) {
+        if (!isEmpty(item.items)) {
+            item.items = item.items.map(it =>
+                buildMenuItemConfig(
+                    it as ChartContextMenuItemLike,
+                    contextMenuEvent,
+                    chartModel,
+                    point,
+                    points
+                )
+            );
+        }
+        if (item.actionFn) {
+            const fn = item.actionFn as ChartMenuItem['actionFn'];
+            item.actionFn = e => {
+                fn(e, contextMenuEvent, {chartModel, point, points});
+            };
+        }
+    }
+
+    return item as MenuItem;
+}
+
+function parseToken(token: string, chartModel: ChartModel): MenuItem | string {
+    switch (token) {
+        case 'viewFullscreen':
+            return {
+                text: 'View in full screen',
+                icon: Icon.expand(),
+                hidden: !chartModel,
```

Comment:
I notice that all of these have this `hidden: !chartModel` - are we ever rendering this menu without a `chartModel`?  Should there be some early out before we reach this point?